robolectric/src/main/java/org/robolectric/android/internal/AndroidTestEnvironment.java

```diff
@@ -692,16 +698,54 @@ private String createTempDir(String name) {
         .toString();
   }
 
+  private static BroadcastReceiver newBroadcastReceiverFromP(
+      String receiverClassName, LoadedApk loadedApk, BroadcastReceiverData receiver) {
+    ClassLoader classLoader = Shadow.class.getClassLoader();
+    if (loadedApk == null || loadedApk.getAppFactory() == null) {
+      return (BroadcastReceiver) newInstanceOf(receiverClassName);
+    } else if (receiver.getActions() == null || receiver.getActions().isEmpty()) {
+      try {
+        return loadedApk.getAppFactory().instantiateReceiver(classLoader, receiverClassName, null);
+      } catch (InstantiationException | IllegalAccessException | ClassNotFoundException ignored) {
+        // Ignore these errors
+      }
+    } else {
+      for (String action : receiver.getActions()) {
```

Comment:
I don't think it's really safe to instantiate arbitrary intents here. If there are some app-specific intents, they may require some specific data or extras that aren't being provided.

I think at this point you always `null` for the Intent.

I would recommend doing removing all the logic that iterates through the intents and tries to instantiate them one-by-one.

Reply:
Robolectric registers a `BroadcastReceiver` instance with the filter that contains all actions that this `BroadcastReceiver` has now. I consider to initialize multiple receiver instances with all actions, but it can't ensure that this `BroadcastReceiver` is instanitated with exact users data, for example they might pass meta data when start this `BroadcastReceiver`. So I decide to use a mid-balance method to try to instantiate this receiver instance by iterating all actions until `AppComponectFactory` accepts one and produce valid receiver instance to us.

Reply:
Does't it seem like the use case for `intent` is in the case where the broadcast receiver is lazy-loaded? For instance, if you send an intent to some arbitrary string action (`MY_ACTION`), and the broadcast receiver for (`MY_ACTION`) is not instantiated, it can then be instantiated with the intent that was sent.

For instance, in this case, the receiver is being instantiated with a specific intent that was sent:
https://cs.android.com/android/platform/superproject/+/master:frameworks/base/core/java/android/app/ActivityThread.java;l=4298-4299;drc=c16c36030a21d6fc21901899d13477c31cc808a7;bpv=1;bpt=1

In AndroidTestEnvironment, there are no intents being sent, so I think we should always be passing `null` as the intent.

Is there some use case that I am missing?

Reply:
> In AndroidTestEnvironment, there are no intents being sent, so I think we should always be passing null as the intent.

Yep. The main reason that I use `Intent` is to avoid receiver's internal state is incorrect when initializing it with null as much as possible. Added tests contain this scenario that `AppComponentFactory` uses different constrcutor data to initialize receiver instance based on different `Intent` action.

Reply:
@hoisie Are there any progresses of this issue?

Reply:
Yeah I still think we should be passing `null` here. This logic is executed at the start of each test , so instantiating arbitrary intents seems misleading.

Perhaps if the BroadcastReceiver instantiation was changed to be lazy we could route the correct intent to the AppCompotentFactory, but receiver instantiation is not lazy at the moment.

Reply:
> Yeah I still think we should be passing null here. This logic is executed at the start of each test , so instantiating receivers with arbitrary intents seems misleading.

Okay. I will update it with `null` tomorrow, and keep this behavior before we find extra issues from users. 

> Perhaps if the BroadcastReceiver instantiation was changed to be lazy we could route the correct intent to the AppCompotentFactory, but receiver instantiation is not lazy at the moment.

We can make receiver instantiation lazy, and I think it is more like the real AOSP logic.

Reply:
@hoisie I updated the logic with `null`. You can check it again.